aws/aws-parallelcluster
# [develop] Add support for Capacity Blocks for ML

### Description of changes
* Added support for `CapacityType=CAPACITY_BLOCK`.
  * To permit the usage of capacity block it's required to set `market_type="capacity-block"` in the Launch Template.
* Default value for `AllocationStrategy` is now set to `None` when using Capacity Blocks

#### Validators
* Extended `CapacityReservationValidator` to verify capacity-type vs reservation-type
  * CapacityReservation of type `capacity-block` have a new attribute `ReservationType`, we're using it to validate that configured `CapacityType` corresponds to it.
* Extended `CapacityReservationValidator` to have a single instance type when using capacity reservation id
   * Capacity Reservation is always associate to a single instance type.
* Extended `ComputeResourceSizeValidator` to validate min count when using CB
  * At the moment all the instances in Capacity Block reservations are managed as static instances, so min count must be equal to max count and this value should be > 0.
  * To perform this check we had to move ComputeResourceSizeValidator registration at queue level since capacity_type is specified at that level.
*Add new `CapacityReservationSizeValidator` to verify number of instances is not exceeded
  * In the cluster config we created a map to keep track of the number of instances associated through max-count to a given capacity reservation id. Then we're verifying that the number of instances is not exceeding the TotalInstanceCount.
  * This will be executed for every capacity reservation, but has a strong value when using Capacity Block reservation.
* Improve CapacityReservationValidator to always validate subnet
  * Previously the subnet was validated only when instance_type was set.

### Additional changes
* [Add missing requirement for test_lambda_handler unit test](https://github.com/aws/aws-parallelcluster/commit/481cccc966794928b07fd067b0acb8172e434dd5)
* Modify describe_capacity_reservations to return a list of CapacityBlockReservationInfo
  * From this moment we don't need to know the structure of the describe-capacity-reservations output in the entire code.  
  * We can use the `CapacityBlockReservationInfo` wrapper class that decouples boto3 structure from parallelcluster internal logic.
  * The code is now more robust and thanks to this change I found a bug in the code: the _validate_no_pg was returning no errors even if there were no capacity reservations without PG. Aligned the unit test for it.

### Tests
* Added unit tests to cover the new/modified code.

#### Manual test of validators

Instance not supporting capacity block: 
```
"type": "CapacityTypeValidator",
"message": "Usage type 'capacity-block' not supported with instance type 't3.large'"
```
Instance type not aligned with CR instance:
```
"type": "ComputeResourceLaunchTemplateValidator",
"message": "Unable to validate configuration parameters for instance type p5.48xlarge. Please double check your cluster configuration. Capacity Reservation's attribute does not match with requested instance parameter. Attribute: InstanceType, requested value: p5.48xlarge, expected value: t3.large"
...
"type": "CapacityReservationValidator",
"message": "Capacity reservation cr-0541d9b022c98faaa must have the same instance type as p5.48xlarge."
```

More instances in the same compute resource:
```
"type": "CapacityReservationSizeValidator",
"message": "Number of instances configured for Capacity Reservation cr-01d2ca74c0d1f1705: 4 is exceeding the total instances count available for the Capacity Reservation: 2"
```

more instances from multiple queues/compute resources, one with 2 and one with 1 from same CB
```
"type": "CapacityReservationSizeValidator",
"message": "Number of instances configured for Capacity Reservation cr-01d2ca74c0d1f1705: 3 is exceeding the total instances count available for the Capacity Reservation: 2"
```

Wrong size: MinCount: 0, MaxCount: 1
```
"type": "ComputeResourceSizeValidator",
"message": "Max count must be set to the same value of min count when using Capacity Block reservation."
...
"type": "ComputeResourceSizeValidator",
"message": "Min count must be a value > 0 when using Capacity Block reservation."
```


### References
* Cookbook changes: https://github.com/aws/aws-parallelcluster-cookbook/pull/2526
* Related patch for the node: https://github.com/aws/aws-parallelcluster-node/pull/591
* Slurm reservations: https://slurm.schedmd.com/reservations.html
* Capacity Blocks: https://docs.aws.amazon.com/en_us/AWSEC2/latest/UserGuide/ec2-capacity-blocks.html
